File: src/providers/together.ts
import type { Provider } from "../core/types.js";

export const together: Provider = {
  name: "together",
  displayName: "Together AI",
  keyEnvVar: "TOGETHER_API_KEY",
  detect: (key) => /^[a-f0-9]{64}$/.test(key) || /^tgp_v\d_[A-Za-z0-9_-]{32,}$/.test(key),
  validate: async (key, { http, signal }) => {
    const res = await http("https://api.together.xyz/v1/models", {
      headers: { Authorization: `Bearer ${key}` },
      signal,
    });
    if (!res.ok) return { status: res.status, ok: false };
    const body = (await res.json().catch(() => null)) as unknown;
    const count = Array.isArray(body)
      ? body.length
      : Array.isArray((body as { data?: unknown[] })?.data)
        ? (body as { data: unknown[] }).data.length
        : 0;
    return {
      status: res.status,
      ok: true,
      metadata: { modelCount: count },
    };
  },
};

File: python/src/llm_key_validator/providers/together.py
from __future__ import annotations

import re

from ..core.types import Provider, ProviderRawResult, ValidationContext

_DETECT_HEX = re.compile(r"^[a-f0-9]{64}$")
_DETECT_TGP = re.compile(r"^tgp_v\d_[A-Za-z0-9_-]{32,}$")


def _detect(key: str) -> bool:
    return bool(_DETECT_HEX.match(key) or _DETECT_TGP.match(key))


def _validate(key: str, ctx: ValidationContext) -> ProviderRawResult:
    res = ctx.http(
        "https://api.together.xyz/v1/models",
        headers={"Authorization": f"Bearer {key}"},
    )
    if not res.ok:
        return ProviderRawResult(status=res.status, ok=False)
    body = res.json()
    if isinstance(body, list):
        count = len(body)
    elif isinstance(body, dict) and isinstance(body.get("data"), list):
        count = len(body["data"])
    else:
        count = 0
    return ProviderRawResult(status=res.status, ok=True, metadata={"modelCount": count})


together = Provider(
    name="together",
    display_name="Together AI",
    key_env_var="TOGETHER_API_KEY",
    detect=_detect,
    validate=_validate,
)
